### PR TITLE
fix(components): fixes issues with tfjob launcher component.

### PR DESCRIPTION
**Pull request by NikeNano**
**Description of your changes:**

When trying to use the sample for the tfjob launcher component, I found the following issues:

- The data types had to be updated from Json to JsonObject
- The workerNum where formatted like strings in the dict -> "3" instead of 3, to fix this a python component where created.

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by NikeNano** [comment]
@Ark-kun could you please take a look?

**Comment by NikeNano** [comment]
Could you take a look at this @numerology?

**Comment by NikeNano** [comment]
Could you maybe take a look at this PR @capri-xiyue, thank you?

File: components/kubeflow/launcher/component.yaml [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
+- {name: PS Spec,                     type: JsonObject, default: '{}',      description: 'TFJob ps replicaSpecs.'}
+- {name: Worker Spec,                 type: JsonObject, default: '{}',      description: 'TFJob worker replicaSpecs.'}
+- {name: Chief Spec,                  type: JsonObject, default: '{}',      description: 'TFJob chief replicaSpecs.'}
+- {name: Evaluator Spec,              type: JsonObject, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
+- {name: Tfjob Timeout Minutes,       type: Integer,    default: 1440,      description: 'Time in minutes to wait for the TFJob to complete.'}
+- {name: Delete Finished Tfjob,       type: Bool,       default: 'True' ,   description: 'Whether to delete the tfjob after it is finished.'}
 implementation:
   container:
-    image: liuhougangxa/kubeflow-tfjob-launcher:latest
+    image: nikenano/launchernew:latest
```
Comment (capri-xiyue) [inline components/kubeflow/launcher/component.yaml:19]: @Bobgy I'm sure whether it's common practise in KFP to use image in personal registry?

Comment (NikeNano) [reply components/kubeflow/launcher/component.yaml:19]: Good point,  I think we should make these images official and keep on `gcr` as is done with other images that are used for Kubeflow pipelines itself.

**Comment by NikeNano** [comment]
Thanks for the review @capri-xiyue

**Comment by Bobgy** [comment]
/cc @kubeflow/wg-training-leads 
Do you want to review? It seems related with tf-job.
Should I cc a subset of people next time?

**Review by gaocegege — COMMENTED** [review]
> The workerNum where formatted like strings in the dict -> "3" instead of 3, to fix this a python component where created.


Can you please explain more about it?

**Comment by NikeNano** [comment]
> > The workerNum where formatted like strings in the dict -> "3" instead of 3, to fix this a python component where created.
> 
> Can you please explain more about it?

@gaocegege  Assuming you replace `JSON` -> `JsonObject` in the component file. You will get the following problem if you run the pipeline:
```bash
INFO:root:Generating tfjob template.
INFO:root:Creating kubeflow.org/tfjobs mnist in namespace kubeflow.
ERROR:root:Exception when create kubeflow.org/tfjobs: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"TFJob.kubeflow.org \"mnist\" is invalid: spec.tfReplicaSpecs.Worker.replicas: Invalid value: \"string\": spec.tfReplicaSpecs.Worker.replicas in body must be of type integer: \"string\"","reason":"Invalid","details":{"name":"mnist","group":"kubeflow.org","kind":"TFJob","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: \"string\": spec.tfReplicaSpecs.Worker.replicas in body must be of type integer: \"string\"","field":"spec.tfReplicaSpecs.Worker.replicas"}]},"code":422}
Traceback (most recent call last):
  File "/ml/launch_tfjob.py", line 136, in <module>
    main()
  File "/ml/launch_tfjob.py", line 126, in main
    create_response = tfjob.create(inst)
  File "/ml/launch_crd.py", line 98, in create
    self._log_and_raise_exception(e, "create")
  File "/ml/launch_crd.py", line 135, in _log_and_raise_exception
    raise ex
kubernetes.client.rest.ApiException: (422)
Reason: Unprocessable Entity
HTTP response headers: HTTPHeaderDict({'Date': 'Fri, 15 Jan 2021 07:30:54 GMT', 'Content-Length': '568', 'Content-Type': 'application/json', 'Cache-Control': 'no-cache, private'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"TFJob.kubeflow.org \"mnist\" is invalid: spec.tfReplicaSpecs.Worker.replicas: Invalid value: \"string\": spec.tfReplicaSpecs.Worker.replicas in body must be of type integer: \"string\"","reason":"Invalid","details":{"name":"mnist","group":"kubeflow.org","kind":"TFJob","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: \"string\": spec.tfReplicaSpecs.Worker.replicas in body must be of type integer: \"string\"","field":"spec.tfReplicaSpecs.Worker.replicas"}]},"code":422}
```

With this being the issue: 

```
"reason":"FieldValueInvalid","message":"Invalid value: \"string\": spec.tfReplicaSpecs.Worker.replicas in body must be of type integer: \"string\"","field":"spec.tfReplicaSpecs.Worker.replicas"
```

I tried different ways to try to convert it to a int inside the operations but the solutions required quite some change, but maybe this is a simpler solution to it :)

**Comment by NikeNano** [comment]
Did the explanation help you @gaocegege?

**Comment by gaocegege** [comment]
Gotcha, SGTM

Sorry for the late reply.

/lgtm

**Comment by NikeNano** [comment]
@Ark-kun or @chensun could you please take a look as well, thanks.

**Comment by NikeNano** [comment]
@Bobgy  is there anyone else we should ask for feedback on this?

**Comment by Bobgy** [comment]
/Approve

/hold
To give
/cc @Ark-kun @chensun @numerology 
a last chance to review.


I think tflauncher component is largely community maintained now.

@gaocegege @NikeNano do you want to add an OWNERS file for it?

**Comment by gaocegege** [comment]
I am glad to add an owner for it. WDYT @NikeNano

**Comment by NikeNano** [comment]
> I am glad to add an owner for it. WDYT @NikeNano

I would be happy to maintain it if that is the question @gaocegege :) ?

**Comment by gaocegege** [comment]
@NikeNano 

Thanks for your contribution! :tada: :+1:

Ref https://github.com/kubeflow/pipelines/pull/5113

**Comment by Bobgy** [comment]
/unhold

Thank you both for the contributions!

**Comment by NikeNano** [comment]
The` kubeflow-pipelines-component-yaml` test seems broken atm.

**Comment by NikeNano** [comment]
@Bobgy is the `kubeflow-pipelines-component-yaml ` test broken atm?

**Comment by NikeNano** [comment]
@capri-xiyue is the kubeflow-pipelines-component-yaml test broken?

**Comment by NikeNano** [comment]
It looks to me like we have some issues that could be related to the v2 of the api, do you know more about this @numerology?
```go
Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "/usr/local/lib/python3.6/site-packages/kfp/__init__.py", line 24, in <module>
    from ._client import Client
  File "/usr/local/lib/python3.6/site-packages/kfp/_client.py", line 31, in <module>
    from kfp.compiler import compiler
  File "/usr/local/lib/python3.6/site-packages/kfp/compiler/__init__.py", line 17, in <module>
    from ..containers._component_builder import build_python_component, build_docker_image, VersionedDependency
  File "/usr/local/lib/python3.6/site-packages/kfp/containers/_component_builder.py", line 32, in <module>
    from kfp.containers import entrypoint
  File "/usr/local/lib/python3.6/site-packages/kfp/containers/entrypoint.py", line 23, in <module>
    from kfp.containers import entrypoint_utils
  File "/usr/local/lib/python3.6/site-packages/kfp/containers/entrypoint_utils.py", line 23, in <module>
    from kfp.pipeline_spec import pipeline_spec_pb2
  File "/usr/local/lib/python3.6/site-packages/kfp/pipeline_spec/pipeline_spec_pb2.py", line 23, in <module>
    create_key=_descriptor._internal_create_key,
AttributeError: module 'google.protobuf.descriptor' has no attribute '_internal_create_key
```
 from https://github.com/kubeflow/pipelines/blob/master/api/v2alpha1/python/setup.py. Do you know anything about this @chensun?

**Comment by chensun** [comment]
> It looks to me like we have some issues that could be related to the v2 of the api, do you know more about this @numerology?
> 
> ```go
> Traceback (most recent call last):
>   File "<string>", line 3, in <module>
>   File "/usr/local/lib/python3.6/site-packages/kfp/__init__.py", line 24, in <module>
>     from ._client import Client
>   File "/usr/local/lib/python3.6/site-packages/kfp/_client.py", line 31, in <module>
>     from kfp.compiler import compiler
>   File "/usr/local/lib/python3.6/site-packages/kfp/compiler/__init__.py", line 17, in <module>
>     from ..containers._component_builder import build_python_component, build_docker_image, VersionedDependency
>   File "/usr/local/lib/python3.6/site-packages/kfp/containers/_component_builder.py", line 32, in <module>
>     from kfp.containers import entrypoint
>   File "/usr/local/lib/python3.6/site-packages/kfp/containers/entrypoint.py", line 23, in <module>
>     from kfp.containers import entrypoint_utils
>   File "/usr/local/lib/python3.6/site-packages/kfp/containers/entrypoint_utils.py", line 23, in <module>
>     from kfp.pipeline_spec import pipeline_spec_pb2
>   File "/usr/local/lib/python3.6/site-packages/kfp/pipeline_spec/pipeline_spec_pb2.py", line 23, in <module>
>     create_key=_descriptor._internal_create_key,
> AttributeError: module 'google.protobuf.descriptor' has no attribute '_internal_create_key
> ```
> 
> from https://github.com/kubeflow/pipelines/blob/master/api/v2alpha1/python/setup.py. Do you know anything about this @chensun?

Yes this error is due to the old version of protobuf being used. I made a recent fix https://github.com/kubeflow/pipelines/pull/5135, but haven't done a new release of kfp-pipeline-spec yet. I'll do a new release today.

**Pull request merged**